test/component_failures_test.py

```diff
@@ -136,7 +143,57 @@ def f(x, j):
     expected_results = 4 * list(range(num_workers_per_scheduler * num_local_schedulers))
     self.assertEqual(results, expected_results)
 
-    ray.worker.cleanup()
+  def check_components_alive(self, component_type, check_component_alive):
+    """Check that a given component type is alive on all worker nodes.
+    """
+    components = ray.services.all_processes[component_type][1:]
+    for component in components:
+      if check_component_alive:
+        self.assertTrue(component.poll() is None)
+      else:
+        self.assertTrue(component.poll() <= 0)
+
+  def testLocalSchedulerFailed(self):
+    # Kill all local schedulers on worker nodes.
+    self._testComponentFailed(ray.services.PROCESS_TYPE_LOCAL_SCHEDULER)
+
+    # The plasma stores and plasma managers should still be alive on the worker
+    # nodes.
+    self.check_components_alive(ray.services.PROCESS_TYPE_PLASMA_STORE, True)
+    self.check_components_alive(ray.services.PROCESS_TYPE_PLASMA_MANAGER, True)
+    self.check_components_alive(ray.services.PROCESS_TYPE_LOCAL_SCHEDULER, False)
+
+  def testPlasmaManagerFailed(self):
+    # Kill all plasma managers on worker nodes.
+    self._testComponentFailed(ray.services.PROCESS_TYPE_PLASMA_MANAGER)
+
+    # The plasma stores should still be alive (but unreachable) on the worker
+    # nodes.
+    self.check_components_alive(ray.services.PROCESS_TYPE_PLASMA_STORE, True)
+    self.check_components_alive(ray.services.PROCESS_TYPE_PLASMA_MANAGER, False)
+    self.check_components_alive(ray.services.PROCESS_TYPE_LOCAL_SCHEDULER, False)
+
+  def testPlasmaStoreFailed(self):
+    # Kill all plasma stores on worker nodes.
+    self._testComponentFailed(ray.services.PROCESS_TYPE_PLASMA_STORE)
+
+    # No processes should be left alive on the worker nodes.
+    self.check_components_alive(ray.services.PROCESS_TYPE_PLASMA_STORE, False)
+    self.check_components_alive(ray.services.PROCESS_TYPE_PLASMA_MANAGER, False)
+    self.check_components_alive(ray.services.PROCESS_TYPE_LOCAL_SCHEDULER, False)
+
+  def testDriverLives(self):
+    ray.worker.init()
+    for process in [
+        ray.services.all_processes[ray.services.PROCESS_TYPE_PLASMA_STORE][0],
+        ray.services.all_processes[ray.services.PROCESS_TYPE_PLASMA_MANAGER][0],
+        ray.services.all_processes[ray.services.PROCESS_TYPE_LOCAL_SCHEDULER][0],
+        ray.services.all_processes[ray.services.PROCESS_TYPE_GLOBAL_SCHEDULER][0],
+        ]:
+      process.terminate()
+      time.sleep(0.1)
+      process.kill()
```

Comment:
why was it necessary to add the `process.kill()`? why wasn't `process.terminate()` enough?